### PR TITLE
Search result component for Booking module

### PR DESCRIPTION
**Pull request by dakatk**

### CODE REVIEW

File: angular/angular.json [hunk 1]
```diff
@@ -20,6 +20,8 @@
             "aot": true,
             "assets": [
               "src/api.account.json",
+              "src/api.booking.json",
+              "src/api.lodging.json",
```
Comment (escalonn) [inline angular/angular.json:24]: lodging doesn't belong here

File: angular/src/api.booking.json [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    "stay": {
+      "id": "1",
+      "checkIn": "2019-01-16",
+      "checkOut": "2019-01-17",
+      "dateCreated": "2019-01-01",
+      "dateModified": ""
+    },
+    "status": "Paid"
+  }
+]
```
Comment (escalonn) [inline angular/src/api.booking.json:71]: this data doesn't match the `Booking[]` type as defined in this branch

File: angular/src/api.lodging.json [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        "id": "1",
+        "accountId": "1",
+        "hotelId": "1",
+        "comment": "Great place!",
+        "date created": "",
+        "rating": "5"
+      }
+    ]
+  }
+]
```
Comment (escalonn) [inline angular/src/api.lodging.json:216]: not in this branch

File: angular/src/app.config.json [hunk 4]
```diff
@@ -19,9 +19,18 @@
...
+        "url": "account"
+      },
+      {
+        "text": "booking",
+        "url": "booking"
+      },
+      {
+        "text": "lodging",
+        "url": "lodging"
+      }
```
Comment (escalonn) [inline angular/src/app.config.json:33]: unnecessary diff. lodging stuff above also

File: angular/src/app/modules/booking/booking/booking.component.ts [hunk 5]
```diff
@@ -1,11 +1,34 @@
...
+  constructor(private lodgingService: LodgingService) {}
+
+  ngOnInit(): void {
+    this.lodgings$ = this.lodgingService.get();
+
+    this.lodgings$.pipe(
+      map((lodgings) => {
+        lodgings.map((lodging) => lodging.rentals);
+      })
+    );
```
Comment (escalonn) [inline angular/src/app/modules/booking/booking/booking.component.ts:25]: what is this code doing? you're not assigning the return value of these pipe calls anywhere

File: angular/src/app/modules/booking/search-results/search-results.component.html [hunk 6]
```diff
@@ -0,0 +1,60 @@
+<section class="hero is-fullheight-with-navbar">
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.html:1]: two-space indent

File: angular/src/app/modules/booking/search-results/search-results.component.html [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                                        Persons • Bedrooms: {{ rental.rentalUnit.bedrooms.length }} • Bathrooms:
+                                        {{ rental.rentalUnit.bathrooms.length }}
+                                        <button class="button is-primary">Reserve Now!</button>
+                                        <br>
+
+                                    </div>
+                                    <p class="discountedPrice" style="margin-left: 2.5rem;"
+                                        *ngIf="!rental.rentalUnit.discountedPrice">${{ rental.rentalUnit.price}}</p>
+                                    <p class="orignalPrice" *ngIf="rental.rentalUnit.discountedPrice">
+                                        ${{ rental.rentalUnit.price}}</p>
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.html:48]: odd spacing

File: angular/tslint.json [hunk 8]
```diff
@@ -1,4 +1,5 @@
 {
+  "strictTemplates": false,
```
Comment (escalonn) [inline angular/tslint.json:2]: this doesn't belong here

File: angular/src/app/modules/booking/booking/booking.component.ts (hunk 5 above)
Comment (dakatk) [reply angular/src/app/modules/booking/booking/booking.component.ts:25]: Had lazy loading issues at first. May have been resolved another way by now, but this code was to keep it consistent with the way that Fred tackled the issued (see account.component.ts, lines 78-80 for reference)

File: angular/src/api.lodging.json (hunk 3 above)
Comment (dakatk) [reply angular/src/api.lodging.json:216]: Based on what Fred said, lodgings appear to be a necessary part of the user searching for a rental to book

File: angular/angular.json (hunk 1 above)
Comment (dakatk) [reply angular/angular.json:24]: From what I understand, we need Lodging info for the search functionality to work correctly

File: angular/src/api.booking.json (hunk 2 above)
Comment (dakatk) [reply angular/src/api.booking.json:71]: What values are allowed for the status field in Booking?

Comment (escalonn) [inline angular/src/api.booking.json:71]: again, this JSON doesn't match `Booking[]` type.

File: angular/src/api.lodging.json [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      }
+    ],
+    "bathrooms": [
+      {
+        "id": "1",
+        "fixture": "1"
+      }
+    ]
+  }
+]
```
Comment (escalonn) [inline angular/src/api.lodging.json:90]: this JSON also does not match `Lodging[]` type.
you can check this by pasting the whole JSON into any TS file after e.g. `const x: Lodging[] =`

File: angular/src/app.config.json (hunk 4 above)
Comment (escalonn) [inline angular/src/app.config.json:33]: this diff is still unnecessary

File: angular/src/app/data/rental-unit.model.ts [hunk 10]
```diff
@@ -15,8 +14,7 @@ import { Bedroom } from './bedroom.model';
  */
 export interface RentalUnit {
   id: string;
-  bathrooms: Bathroom[];
-  bedrooms: Bedroom[];
+  bedrooms: Bedroom;
```
Comment (escalonn) [inline angular/src/app/data/rental-unit.model.ts:17]: "bedrooms" is one bedroom?
didn't we talk about taking out bedrooms entirely? or was it this - i dont remember

File: angular/src/app/data/rental.model.ts [hunk 11]
```diff
@@ -12,5 +12,8 @@ import { RentalUnit } from './rental-unit.model';
 export interface Rental {
   id: string;
   name: string;
+  status: string;
+  price: number;
+  discountedPrice?: number;
```
Comment (escalonn) [inline angular/src/app/data/rental.model.ts:17]: no documentation on the meaning here, i would at least expect documentation in the PR description

File: angular/src/app/modules/booking/booking/booking.component.html [hunk 12]
```diff
@@ -108,4 +111,4 @@ <h1 class="title is-2">booking</h1>
       </div>
     </div>
   </div>
-</section>
+</section>
```
Comment (escalonn) [inline angular/src/app/modules/booking/booking/booking.component.html:114]: only the `class` additions here are meaningful diffs; the rest of the project is not using xml-style self-closing tags

File: angular/src/app/modules/booking/booking/booking.component.spec.ts [hunk 13]
```diff
@@ -1,13 +1,77 @@
...
+import { LodgingService } from 'src/app/services/lodging/lodging.service';
+import { of } from 'rxjs';
+import { SearchResultsComponent } from '../search-results/search-results.component';
 
 describe('BookingComponent', () => {
   let component: BookingComponent;
   let fixture: ComponentFixture<BookingComponent>;
 
+  const lodgingServiceStub = {
+    get() {
```
Comment (escalonn) [inline angular/src/app/modules/booking/booking/booking.component.spec.ts:12]: add return type annotation

File: angular/src/app/modules/booking/search-results/search-results.component.html [hunk 14]
```diff
@@ -0,0 +1,60 @@
+<section class="hero is-fullheight-with-navbar">
+  <div class="hero-head">
+    <div class="container">
+      <br>
+      <h3 class="title is-2">search results for "{{query}}"</h3>
+      <div *ngIf="lodgings !== null">
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.html:6]: just have `"lodgings"`; e.g. undefined would also be a problem

File: angular/src/app/modules/booking/search-results/search-results.component.html [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+                            </span>
+                          </span>
+                        </div>
+                        <p class="subtitle is-6">Short Description</p>
+                      </div>
+                    </div>
+
+                    <div class="content">
+                      {{ rental.rentalUnit.type }} • Occupancy: {{ rental.rentalUnit.occupancy }}
+                      Persons • Bedrooms: {{ rental.rentalUnit.bedrooms.count }}
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.html:39]: break lines a little more sensibly

File: angular/src/app/modules/booking/search-results/search-results.component.html [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+                  <div class="card-content">
+                    <div class="media">
+
+                      <div class="media-content">
+                        <div class="title is-4">
+                          <span>{{ rental.name }}</span>
+                          <span class="end-of-card stars">
+                            <span *ngFor="let star of averageRating(lodging)">
+                              <p *ngIf="star" class="fa fa-star checked"></p>
+                              <p *ngIf="!star" class="fa fa-star unchecked"></p>
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.html:29]: it's wrong to put a p inside a span. also fyi you can use class binding for this. e.g.
`<span *ngFor="let star of averageRating(lodging)" class="fa fa-star" [class]="{checked: star, unchecked: !star}">`
you could consider just having one "checked" class here which toggles the appearance in itself

File: angular/src/app/modules/booking/search-results/search-results.component.html [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+                    <div class="content">
+                      {{ rental.rentalUnit.type }} • Occupancy: {{ rental.rentalUnit.occupancy }}
+                      Persons • Bedrooms: {{ rental.rentalUnit.bedrooms.count }}
+                      <button class="button is-primary">Reserve Now!</button>
+                      <br>
+
+                    </div>
+                    <p class="discountedPrice" style="margin-left: 2.5rem;" *ngIf="!rental.discountedPrice">
+                      ${{ rental.price}}</p>
+                    <p class="orignalPrice" *ngIf="rental.discountedPrice">
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.html:46]: spelling in class name (here and in css)
also weird spacing in the interpolations here as well as weird line breaks

File: angular/src/app/modules/booking/search-results/search-results.component.spec.ts [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+          status: 'available',
+        },
+      ],
+      reviews: [
+        {
+          id: '1',
+          accountId: '1',
+          hotelId: '1',
+          comment: 'comment',
+          dateCreated: new Date(Date.now()),
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.spec.ts:50]: this is redundant

File: angular/src/app/modules/booking/search-results/search-results.component.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+import { Lodging } from 'src/app/data/lodging.model';
+
+@Component({
+  selector: 'uic-search-results',
+  templateUrl: './search-results.component.html',
+  styleUrls: ['./search-results.component.scss'],
+})
+export class SearchResultsComponent implements OnInit {
+  @Input() lodgings: Lodging[] | null;
+  query = 'test query';
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.ts:11]: test query for master branch?

File: angular/tsconfig.json [hunk 20]
```diff
@@ -17,10 +17,10 @@
     "sourceMap": true,
     "strictNullChecks": true,
     "target": "es2015",
-    "lib": ["es2018", "dom"],
+    "lib": ["es2018", "dom"]
   },
   "angularCompilerOptions": {
     "strictInjectionParameters": true,
-    "strictTemplates": true,
+    "strictTemplates": true
```
Comment (escalonn) [inline angular/tsconfig.json:24]: prettier violation, and unneeded diff anyway

File: angular/src/app/modules/booking/search-results/search-results.component.spec.ts (hunk 18 above)
Comment (dakatk) [reply angular/src/app/modules/booking/search-results/search-results.component.spec.ts:50]: Date.now() returns a number

File: angular/src/app/modules/booking/booking/booking.component.ts [hunk 21]
```diff
@@ -16,5 +18,23 @@ export class BookingComponent implements OnInit {
...
+      )
+    }
+
+    this.lodgings$.pipe(
+      map(
+        (lodgings) => lodgings.map(
+          (lodging) => lodging.reviews = mapDates(lodging.reviews)
+        )
+      )
+    )
```
Comment (escalonn) [inline angular/src/app/modules/booking/booking/booking.component.ts]: my rxjs skills are low but this looks wrong in several ways. isn't this pipe doing nothing, and needs to be part of the statement on line 20, so that the property is set to the return value of pipe? the indents are also odd, and i think this return value would be something unreasonable (the reviews only, and not the lodgings)

if you want to do this kind of thing then you should have additional new types that contain Dates instead of strings. the alternative is transforming the date only at the point of display.

missing semicolon, probably other prettier issues.

File: angular/src/app/modules/booking/search-results/search-results.component.html [hunk 22]
```diff
@@ -24,10 +24,8 @@ <h3>number of results found: {{ lodgings.length }}</h3>
                         <div class="title is-4">
                           <span>{{ rental.name }}</span>
                           <span class="end-of-card stars">
-                            <span *ngFor="let star of averageRating(lodging)">
-                              <p *ngIf="star" class="fa fa-star checked"></p>
-                              <p *ngIf="!star" class="fa fa-star unchecked"></p>
-                            </span>
+                            <span *ngFor="let star of averageRating(lodging)" class="fa fa-star"
+                              [class]="{checked: star}"></span>
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.html:28]: if there's just the one class, then the nicer syntax is `[class.checked]="star"`

File: angular/src/app/modules/booking/booking/booking.component.ts [hunk 23]
```diff
@@ -16,5 +18,23 @@ export class BookingComponent implements OnInit {
 
   ngOnInit(): void {
     this.lodgings$ = this.lodgingService.get();
+
+    function mapDates(reviews: Review[]): Review[] {
+      return reviews.map(
+        (review) => {
+
+          review.dateCreated = new Date(review.dateCreated);
```
[the remark this answers is not in the document]
Comment (dakatk) [reply angular/src/app/modules/booking/booking/booking.component.ts]: If we had them kept as dates, this is actually the correct way in typescript. If you were to run 'console.log(typeof (review.dateCreated))' before mapping it to a date, it would return 'string'. A type annotation is not the data's actual guaranteed type. This is why Linting exists. A value annotated with 'Date' can be set to a 'string', which is why this mapping would be necessary

File: angular/src/app/modules/booking/search-results/search-results.component.spec.ts [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
```
Comment (escalonn) [inline angular/src/app/modules/booking/search-results/search-results.component.spec.ts:50]: yeah but the zero-arg `new Date()` does the same thing. also there's a prettier issue in this file
